zerver/lib/import_realm.py

```diff
@@ -2080,3 +2101,62 @@ def add_users_to_system_user_groups(
         )
         for membership in usergroup_memberships
     )
+
+
+ZULIP_CLOUD_ONLY_APP_NAMES = ["zilencer", "corporate"]
+
+
+def check_migration_status(exported_migration_status: MigrationStatusJson) -> None:
+    mismatched_migrations_log: dict[str, str] = {}
+    local_migration_status = MigrationStatusJson(
+        migrations_by_app=get_migrations_by_app(), zulip_version=ZULIP_VERSION
+    )
+
+    if exported_migration_status["zulip_version"] != local_migration_status["zulip_version"]:
```

Comment:
I think this is going to be different for most systems running a `Git` version, even if it's one commit different that has no impact on migrations. In particular, I expect this to always fail when trying to import a Zulip Cloud export into a self-hosted system, because the exact commits IDs will often be different. 

Maybe the right answer is to do a split on these values before comparison.

Reply:
```diff
diff --git a/zerver/lib/import_realm.py b/zerver/lib/import_realm.py
index 6ec40763bb..eb349ab258 100644
--- a/zerver/lib/import_realm.py
+++ b/zerver/lib/import_realm.py
@@ -2112,11 +2112,20 @@ def check_migration_status(exported_migration_status: MigrationStatusJson) -> No
         migrations_by_app=get_migrations_by_app(), zulip_version=ZULIP_VERSION
     )
 
-    if exported_migration_status["zulip_version"] != local_migration_status["zulip_version"]:
+    # Different major versions are the most common form of mismatch
+    # and should get a nice error message focused on that, not
+    # migrations.
+    #
+    # We could split on `-`, to get the maintenance release version,
+    # but unless migrations are different, it should generally be safe
+    # to import across minor release differences.
+    exported_primary_version = exported_migration_status["zulip_version"].split(".")[0]
+    local_primary_version = local_migration_status["zulip_version"].split(".")[0]
+    if exported_primary_version != local_primary_version:
         raise Exception(
-            "Zulip version mismatch.\n"
+            "Export was generated on a different Zulip major version.\n"
             f"Export={exported_migration_status['zulip_version']}\n"
-            f"Local={local_migration_status['zulip_version']}"
+            f"Server={local_migration_status['zulip_version']}"
         )
     exported_migrations_by_app = exported_migration_status["migrations_by_app"]
     local_migrations_by_app = local_migration_status["migrations_by_app"]
@@ -2129,16 +2138,12 @@ def check_migration_status(exported_migration_status: MigrationStatusJson) -> No
         if app in ZULIP_CLOUD_ONLY_APP_NAMES and (
             local_app_migrations is None or exported_app_migrations is None
         ):
-            # In the case of a migration from Zulip Cloud, where one
-            # realm has zilencer/corporate apps installed but the
-            # other (self-hosted) server does not, we do not inform
-            # the user of mismatches in these installed apps.
+            # This applications are expected to be present only on
+            # Zulip Cloud, so don't warn about them.
             continue
 
         if not exported_app_migrations:
-            logging.warning(
-                "Local server has '%s' app installed, but exported realm does not.", app
-            )
+            logging.warning("This server has '%s' app installed, but exported realm does not.", app)
         elif not local_app_migrations:
             logging.warning("Exported realm has '%s' app installed, but this server does not.", app)
         elif local_app_migrations != exported_app_migrations:
@@ -2155,8 +2160,13 @@ def check_migration_status(exported_migration_status: MigrationStatusJson) -> No
         ]
 
         error_message = (
-            "Migration mismatch between the exported realm and the local server; showing diff.\n"
+            "Export was generated on a different Zulip version.\n"
+            f"Export={exported_migration_status['zulip_version']}\n"
+            f"Server={local_migration_status['zulip_version']}\n"
+            "\n"
+            "Database foramts differ between the exported realm and this server.\n"
+            "Printing migrations that differ between the versions:\n"
             "--- exported realm\n"
-            "+++ local server"
+            "+++ this server"
         ) + "\n".join(sorted_error_log)
         raise Exception(error_message)
diff --git a/zerver/tests/fixtures/import_fixtures/check_migrations_errors/extra_migrations_error.txt b/zerver/tests/fixtures/import_fixtures/check_migrations_errors/extra_migrations_error.txt
index aa3bb99282..b7f9d07e14 100644
--- a/zerver/tests/fixtures/import_fixtures/check_migrations_errors/extra_migrations_error.txt
+++ b/zerver/tests/fixtures/import_fixtures/check_migrations_errors/extra_migrations_error.txt
@@ -1,6 +1,11 @@
-Migration mismatch between the exported realm and the local server; showing diff.
+Export was generated on a different Zulip version.
+Export=10.0-dev+git
+Server=10.0-dev+git
+
+Database foramts differ between the exported realm and this server.
+Printing migrations that differ between the versions:
 --- exported realm
-+++ local server
++++ this server
 'auth' app:
  0002_alter_permission_name_max_length
 -0003_alter_user_email_max_length
diff --git a/zerver/tests/fixtures/import_fixtures/check_migrations_errors/unapplied_migrations_error.txt b/zerver/tests/fixtures/import_fixtures/check_migrations_errors/unapplied_migrations_error.txt
index 8e089c5eb0..14ad2fdaef 100644
--- a/zerver/tests/fixtures/import_fixtures/check_migrations_errors/unapplied_migrations_error.txt
+++ b/zerver/tests/fixtures/import_fixtures/check_migrations_errors/unapplied_migrations_error.txt
@@ -1,6 +1,11 @@
-Migration mismatch between the exported realm and the local server; showing diff.
+Export was generated on a different Zulip version.
+Export=10.0-dev+git
+Server=10.0-dev+git
+
+Database foramts differ between the exported realm and this server.
+Printing migrations that differ between the versions:
 --- exported realm
-+++ local server
++++ this server
 'auth' app:
  0002_alter_permission_name_max_length
 +0003_alter_user_email_max_length
diff --git a/zerver/tests/test_import_export.py b/zerver/tests/test_import_export.py
index 2567c11546..d0910a3dcd 100644
--- a/zerver/tests/test_import_export.py
+++ b/zerver/tests/test_import_export.py
@@ -2136,8 +2136,8 @@ class RealmImportExportTest(ExportFile):
             )
             do_import_realm(get_output_dir(), "test-zulip")
         missing_apps_log = [
-            "WARNING:root:Local server has 'phonenumber' app installed, but exported realm does not.",
-            "WARNING:root:Local server has 'sessions' app installed, but exported realm does not.",
+            "WARNING:root:This server has 'phonenumber' app installed, but exported realm does not.",
+            "WARNING:root:This server has 'sessions' app installed, but exported realm does not.",
         ]
         self.assertEqual(sorted(mock_log.output), sorted(missing_apps_log))
         self.assertTrue(Realm.objects.filter(string_id="test-zulip").exists())
@@ -2201,7 +2201,11 @@ class RealmImportExportTest(ExportFile):
                 get_output_dir(),
                 "test-zulip",
             )
-        expected_error_message = "Zulip version mismatch.\n" f"Export={ZULIP_VERSION}\n" "Local=8.0"
+        expected_error_message = (
+            "Export was generated on a different Zulip major version.\n"
+            f"Export={ZULIP_VERSION}\n"
+            "Server=8.0"
+        )
         self.assertEqual(expected_error_message, str(e.exception))
 
 
```